inbm/dispatcher-agent/dispatcher/sota/downloader.py

```diff
@@ -165,3 +168,70 @@ def download(self,
 
     def check_release_date(self, release_date: Optional[str]) -> bool:
         return self.is_valid_release_date(release_date)
+
+
+class TiberOSDownloader(Downloader):
+    """TiberOSDownloader class, child of Downloader
+
+       @param signature: signature used to preform signature check on downloaded image.
+    """
+
+    def __init__(self, signature: Optional[str] = None) -> None:
+        super().__init__()
+        self._signature = signature
+
+    def download(self,
+                 dispatcher_broker: DispatcherBroker,
+                 uri: Optional[CanonicalUri],
+                 repo: IRepo,
+                 username: Optional[str],
+                 password: Optional[str],
+                 release_date: Optional[str]) -> None:
+        """Downloads files and places image in local cache
+
+        @param dispatcher_broker: DispatcherBroker object used to communicate with other INBM services
+        @param uri: URI of the source location
+        @param repo: repository for holding the download
+        @param username: username to use for download
+        @param password: password to use for download
+        @param release_date: manifest release date
+        @raises SotaError: release date is not valid
+        """
+
+        if uri is None:
+            raise SotaError("URI is None while performing TiberOS download")
+
+        password = read_oras_token()
+
+        oras_download(dispatcher_broker=dispatcher_broker,
+                      uri=uri,
+                      repo=repo,
+                      umask=UMASK_OTA,
+                      username=username,
+                      password=password)
+
+        # Perform signature check.
+        # Multiple files may have been downloaded from OCI.
+        # The method below will iterate over all files in the repo, calculate the SHA256sum for each file,
+        # and compare it with the provided signature.
+        try:
+            if self._signature:
+                logger.debug("Perform signature check on the downloaded file.")
+                dest_repo = repo.get_repo_path()
+                for filename in os.listdir(dest_repo):
+                    filepath = os.path.join(dest_repo, filename)
+                    if os.path.isfile(filepath):
+                        with open(filepath, 'rb') as file:
+                            file_checksum = hashlib.sha256(file.read()).hexdigest()
+                            if file_checksum == self._signature:
+                                return
```

Comment:
question here: if we discover the file that matches the signature, how does the caller of `download` know which one is matching? (as we need to ensure that the caller is being given a file that matches the signature, and we are downloading multiple files)

Reply:
Based on the sequence diagram, the INBM will pass the local path to UT. I think UT will iterate through the files and find the raw image file.

![image](https://github.com/user-attachments/assets/c38a0576-a5bf-4fd3-87ef-f6581db03e47)

Reply:
@pillaiabhishek can you comment here on UT design? my concern is this could allow a replay attack (see my IM in Teams)

Reply:
Both INBM and UT should not be iterating through the list of files. I am assuming here that there is only one raw image file and the files getting downloaded for the a specific release-tag from oras have different extensions. 
INBM should be using the signature provided by MM to perform the integrity checks of the downloaded raw image file and then provide that file path as an input to UT.

I do anticipate a few rounds of discussion with the BSP, Release, MM teams to nail it down to which exact file  , path and signature details.

Reply:
@yengliong93 Can you add a Jira ticket for this, and a FIXME - (reference to ticket)?

Reply:
JIRA ticket filed - https://jira.devtools.intel.com/browse/NEXMANAGE-734